File: 0731-my-calendar-ii/0731-my-calendar-ii.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

/**
 * @struct MyCalendarTwo
 * @brief Struct to hold the booked intervals and manage double bookings.
 *
 * - single_booked: A dynamic array of intervals representing single bookings.
 * - double_booked: A dynamic array of intervals representing double bookings.
 * - single_size: Current size of the single_booked array.
 * - double_size: Current size of the double_booked array.
 * - single_capacity: Current capacity of the single_booked array.
 * - double_capacity: Current capacity of the double_booked array.
 */
typedef struct {
    int* single_booked;    // Array to store single bookings in [start1, end1, start2, end2,...] format
    int* double_booked;    // Array to store double bookings in the same format
    int single_size;       // Number of elements (intervals) in single_booked
    int double_size;       // Number of elements (intervals) in double_booked
    int single_capacity;   // Current capacity of single_booked array
    int double_capacity;   // Current capacity of double_booked array
} MyCalendarTwo;

/**
 * @brief Creates and initializes a new MyCalendarTwo object.
 * 
 * @return MyCalendarTwo* Pointer to the created calendar object.
 * Initializes the arrays with a starting capacity of 10 intervals.
 */
MyCalendarTwo* myCalendarTwoCreate() {
    MyCalendarTwo* obj = (MyCalendarTwo*)malloc(sizeof(MyCalendarTwo));
    obj->single_capacity = 10; // Initial capacity for single_booked intervals
    obj->double_capacity = 10; // Initial capacity for double_booked intervals
    obj->single_size = 0;      // Initially no single bookings
    obj->double_size = 0;      // Initially no double bookings
    // Allocate memory for both booking arrays
    obj->single_booked = (int*)malloc(sizeof(int) * obj->single_capacity * 2); // Stores intervals [start1, end1, start2, end2,...]
    obj->double_booked = (int*)malloc(sizeof(int) * obj->double_capacity * 2); // Same format for double bookings
    return obj;
}

/**
 * @brief Dynamically grows the given array if necessary.
 * 
 * @param arr Pointer to the array that needs to be resized.
 * @param size Current size of the array.
 * @param capacity Current capacity of the array.
 * @return int* Pointer to the resized array if the capacity was increased.
 * This function doubles the capacity if the size reaches the current capacity.
 */
int* growArrayIfNeeded(int* arr, int* size, int* capacity) {
    if (*size >= *capacity) {
        *capacity *= 2;  // Double the capacity
        arr = (int*)realloc(arr, sizeof(int) * (*capacity) * 2); // Reallocate with new capacity
    }
    return arr;
}

/**
 * @brief Checks for overlaps between existing intervals in the array and a new interval [start, end).
 * 
 * @param arr Pointer to the intervals array.
 * @param size Number of elements in the intervals array.
 * @param start Start time of the new interval.
 * @param end End time of the new interval.
 * @return true if there is an overlap, false otherwise.
 * Checks whether any intervals in the array overlap with the new event's interval [start, end).
 */
bool hasOverlap(int* arr, int size, int start, int end) {
    for (int i = 0; i < size; i += 2) {
        // Check if the new interval [start, end) overlaps with the existing interval arr[i, i+1)
        if (start < arr[i + 1] && end > arr[i]) {  // Overlapping condition
            return true;
        }
    }
    return false;
}

/**
 * @brief Adds a new interval to the given intervals array.
 * 
 * @param arr Pointer to the intervals array.
 * @param size Pointer to the current number of elements in the array.
 * @param capacity Pointer to the capacity of the array.
 * @param start Start time of the new interval.
 * @param end End time of the new interval.
 * 
 * Dynamically grows the array if necessary and appends the new interval to the array.
 */
void addInterval(int** arr, int* size, int* capacity, int start, int end) {
    // Ensure the array has enough capacity to add a new interval
    *arr = growArrayIfNeeded(*arr, size, capacity);
    (*arr)[*size] = start;       // Add start time
    (*arr)[*size + 1] = end;     // Add end time
    *size += 2;                  // Increase the size by 2 to account for the new interval
}

/**
 * @brief Attempts to book a new event in the calendar.
 * 
 * @param obj Pointer to the MyCalendarTwo object.
 * @param start Start time of the event.
 * @param end End time of the event.
 * @return true if the booking is successful and does not cause a triple booking, false otherwise.
 * 
 * First checks if the new event would overlap with any double-booked intervals (which would cause a triple booking).
 * If no triple booking is detected, it adds overlapping intervals from the single bookings to double_booked.
 * Finally, the new event is added to the single_booked array.
 */
bool myCalendarTwoBook(MyCalendarTwo* obj, int start, int end) {
    // Check if there is an overlap with the double-booked intervals (triple booking)
    if (hasOverlap(obj->double_booked, obj->double_size, start, end)) {
        return false;
    }

    // Find overlaps with single-booked intervals and add the overlapping parts to double_booked
    for (int i = 0; i < obj->single_size; i += 2) {
        if (start < obj->single_booked[i + 1] && end > obj->single_booked[i]) {  // Overlapping condition
            int overlap_start = (start > obj->single_booked[i]) ? start : obj->single_booked[i];
            int overlap_end = (end < obj->single_booked[i + 1]) ? end : obj->single_booked[i + 1];
            // Add the overlapping interval to the double_booked list
            addInterval(&(obj->double_booked), &(obj->double_size), &(obj->double_capacity), overlap_start, overlap_end);
        }
    }

    // Add the new event to the single_booked intervals
    addInterval(&(obj->single_booked), &(obj->single_size), &(obj->single_capacity), start, end);
    return true;
}

/**
 * @brief Frees the allocated memory for the MyCalendarTwo object.
 * 
 * @param obj Pointer to the MyCalendarTwo object.
 * Frees memory for both the single_booked and double_booked arrays, as well as the MyCalendarTwo object itself.
 */
void myCalendarTwoFree(MyCalendarTwo* obj) {
    free(obj->single_booked);  // Free memory for single_booked array
    free(obj->double_booked);  // Free memory for double_booked array
    free(obj);                 // Free the MyCalendarTwo object
}


/**
 * Your MyCalendarTwo struct will be instantiated and called as such:
 * MyCalendarTwo* obj = myCalendarTwoCreate();
 * bool param_1 = myCalendarTwoBook(obj, start, end);
 * 
 * myCalendarTwoFree(obj);
 */

File: Daily Questions/0731-my-calendar-ii/0731-my-calendar-ii.py
import bisect

class MyCalendarTwo:
    """
    A calendar system that allows event bookings and prevents triple bookings.

    The MyCalendarTwo class allows users to book events on a calendar, ensuring
    that no three events overlap at any given time.
    """
    
    def __init__(self):
        """
        Initializes the MyCalendarTwo object with two lists:
        
        - single_booked: Stores intervals that represent single bookings.
        - double_booked: Stores intervals where double bookings have occurred.
        """
        self.single_booked = []  # Stores single bookings in sorted order
        self.double_booked = []  # Stores double bookings in sorted order

    def get_overlapping_intervals(self, intervals, start, end):
        """
        Finds the overlapping intervals in the given intervals list with the range [start, end).

        Parameters:
        intervals (list): A list of intervals (flattened) that represent bookings.
        start (int): The start time of the new booking.
        end (int): The end time of the new booking.

        Returns:
        list: A list of overlapping intervals.
        """
        left_index = bisect.bisect_left(intervals, start)
        right_index = bisect.bisect_right(intervals, end)
        
        overlapping_intervals = []
        
        # Handle the case where the left bound is within an interval
        if left_index % 2:
            if intervals[left_index] != start:
                overlapping_intervals.append(start)
            else:
                left_index += 1

        overlapping_intervals += intervals[left_index:right_index]

        # Handle the case where the right bound is within an interval
        if right_index % 2:
            if intervals[right_index - 1] != end:
                overlapping_intervals.append(end)
            else:
                overlapping_intervals.pop()

        return overlapping_intervals
        
    def add_interval(self, intervals, start, end):
        """
        Adds the new interval [start, end) to the given intervals list, maintaining the order.

        Parameters:
        intervals (list): A list of intervals (flattened) that represent bookings.
        start (int): The start time of the new booking.
        end (int): The end time of the new booking.
        """
        left_index = bisect.bisect_left(intervals, start)
        right_index = bisect.bisect_right(intervals, end)
        
        new_intervals = []
        
        # Add the start boundary if it's outside of any existing interval
        if left_index % 2 == 0:
            new_intervals.append(start)
            
        # Add the end boundary if it's outside of any existing interval
        if right_index % 2 == 0:
            new_intervals.append(end)

        # Replace the affected range with the new intervals
        intervals[left_index:right_index] = new_intervals

    def book(self, start: int, end: int) -> bool:
        """
        Attempts to book a new event in the calendar.

        The function checks if the new event overlaps with any double-booked
        intervals. If it does, the booking is rejected to prevent a triple booking.
        If the booking only overlaps with single bookings, it is allowed, and any
        new double-booking intervals are added to the double_booked list.

        Parameters:
        start (int): The start time of the event (inclusive).
        end (int): The end time of the event (exclusive).

        Returns:
        bool: Returns True if the booking is successful and does not cause a triple
              booking. Returns False if the booking would result in a triple booking.
        """
        # Check if the new booking would result in a triple booking
        if self.get_overlapping_intervals(self.double_booked, start, end):
            return False
        
        # Find overlaps with single bookings
        overlapping_single_intervals = self.get_overlapping_intervals(self.single_booked, start, end)

        # If there's an overlap with single bookings, add those overlapping intervals to double_booked
        if overlapping_single_intervals:
            for i in range(0, len(overlapping_single_intervals), 2):
                overlap_start = overlapping_single_intervals[i]
                overlap_end = overlapping_single_intervals[i + 1]
                self.add_interval(self.double_booked, overlap_start, overlap_end)

        # Add the new event to single_booked
        self.add_interval(self.single_booked, start, end)

        return True


# Your MyCalendarTwo object will be instantiated and called as such:
# obj = MyCalendarTwo()
# param_1 = obj.book(start,end)